### PR TITLE
Speed up domain propagation

### PR DESCRIPTION
**Pull request by arcondello**
Make `PresolveImpl::domain_propagation()` `O(num_constraints*num_variables)` rather than `O(num_constraints*num_variables*num_variables)` at the cost of an additional magic number and a more complicated implementation.

As noted previously https://github.com/dwavesystems/dwave-preprocessing/blob/94efdabaca4d030d12a14c10c05319c93362f3cd/dwave/preprocessing/src/presolveimpl.hpp#L570-L577 there was a faster implementation of domain propagation but it created a lot more complexity in the function. Originally we had opted to keep the slower implementation, but testing shows that it is a huge performance bottleneck. So I went ahead and implemented the faster method.

### CODE REVIEW

File: dwave/preprocessing/src/presolveimpl.hpp [hunk 1]
```diff
@@ -928,6 +851,123 @@ class PresolverImpl {
...
+            return false;  // no changes
+        }
+
+        // In order to not fall into zeno's paradox here, we don't shrink bounds when
+        // the reduction is small.
+        static constexpr double BOUND_CHANGE_MINIMUM = 1.0e-3 * FEASIBILITY_TOLERANCE;
+
+        // Very large activities will create numeric issues. So we don't do domain
+        // propagation on those
+        static constexpr double MAX_ACTIVITY = 1.0e10;
```
Comment (elham-azd) [inline dwave/preprocessing/src/presolveimpl.hpp:878]: Could you tell more about the reason behind choosing 1.0e10 as MAX_ACTIVITY? Why not a larger number like 1.0e30 which is INF

Comment (elham-azd) [inline dwave/preprocessing/src/presolveimpl.hpp:878]: Could you tell more about the reason behind choosing `1.0e10` as `MAX_ACTIVITY`? Why not a larger number?

File: dwave/preprocessing/src/presolveimpl.hpp [hunk 2]
```diff
@@ -928,6 +851,123 @@ class PresolverImpl {
...
+                }
+
+                // Since we've marked it as large, let's zero it out to make
+                // some of the sums later a bit safer
+                activities.back() = 0;
+            }
+        }
+
+        // Get the total activity of everything (except the large activity variables)
+        const bias_type activity =
```
Comment (elham-azd) [inline dwave/preprocessing/src/presolveimpl.hpp:913]: Could use `std::accumulate()` for more readability, or calculate activities using a prefix sum in the previous loop as we visit each variable. But it shouldn't make a huge difference anyway.

File: dwave/preprocessing/src/presolveimpl.hpp (hunk 1 above)
Comment (arcondello) [reply dwave/preprocessing/src/presolveimpl.hpp:878]: To be honest it's a bit arbitrary. However, I did run it with both `1e10` and `1e15` on all of the problems in [MQLIB](https://github.com/MQLib/MQLib) and the higher bound did not find any more variables than `1e10` did. So I went with the safer number.

**Pull request merged**